File: JAVA/14_Backtracking/BOJ_2422_한윤정이이탈리아에가서아이스크림을사먹는데.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int n, m, cnt = 0;
    static int[] arr = new int[3];
    static boolean[][] mixed;

    public static void dfs(int dpth, int idx) {
        if (dpth == 3) {
            for (int i = 0; i < 3; i++) {
                if (mixed[arr[i]][arr[(i + 1) % 3]]) {
                    return;
                }
            }
            cnt++;
            return;
        }
        for (int i = idx; i <= n; i++) {
            arr[dpth] = i;
            dfs(dpth + 1, i + 1);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        mixed = new boolean[n + 1][n + 1];
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            mixed[x][y] = mixed[y][x] = true;
        }
        dfs(0, 1);
        System.out.println(cnt);
    }
}

File: C/14_Backtracking/BOJ_2422_한윤정이이탈리아에가서아이스크림을사먹는데.c
#include <stdio.h>
int n, m, arr[3], **mixed, cnt = 0;
void dfs(int dpth, int idx) {
	if (dpth == 3) {
		for (int i = 0; i < 3; i++) {
			if (mixed[arr[i]][arr[(i + 1) % 3]]) {
				return;
			}
		}
		cnt++;
		return;
	}
	for (int i = idx; i <= n; i++) {
		arr[dpth] = i;
		dfs(dpth + 1, i + 1);
	}
}
main() {
	scanf("%d %d", &n, &m);
	mixed = (int**)malloc(sizeof(int*) * (n + 1));
	for (int i = 0; i <= n; i++) {
		mixed[i] = (int*)malloc(sizeof(int) * (n + 1));
		for (int j = 0; j <= n; j++) {
			mixed[i][j] = 0;
		}
	}
	for (int i = 0; i < m; i++) {
		int x, y;
		scanf("%d %d", &x, &y);
		mixed[x][y] = mixed[y][x] = 1;
	}
	dfs(0, 1);
	printf("%d", cnt);
}